# SplitFusion - a fast pipeline for detection of gene fusion based on fusion-supporting split alignment.

Gene fusion is a hallmark of cancer. Many gene fusions are effective therapeutic targets such as BCR-ABL in chronic myeloid leukemia, EML4-ALK in lung cancer, and any of a number of partners-ROS1 in lung cancer. Accurate detection of gene fusion plays a pivotal role in precision medicine by matching the right drugs to the right patients.

Challenges in the diagnosis of gene fusions include poor sample quality, limited amount of available clinical specimens, and complicated gene rearrangements. The anchored multiplex PCR (AMP) is a clinically proven technology designed, in one purpose, for robust detection of gene fusions across clinical samples of different types and varied qualities, including RNA extracted from FFPE samples.

**SplitFusion** is a companion data pipeline for AMP, for the detection of gene fusion based on split alignments, i.e. reads crossing fusion breakpoints, with the ability to accurately infer in-frame or out-of-frame of fusion partners of a given fusion candidate. SplitFusion also outputs example breakpoint-supporting seqeunces in FASTA format, allowing for further investigations.

## Reference publication
[Zheng Z, et al. Split multiplex PCR for targeted next-generation sequencing. Nat Med. 2014](http://www.nature.com/nm/journal/v20/n12/full/nm.3729.html)

## How does SplitFusion work?  


The analysis consists of ## computational steps:

1. Retrive all alignments that have secondary alignments (the 'SA' tag in SAM format) from bam files generated by BWA MEM.

2. Remove alignments with low mapping quality (default 20).

3. The consolidated SA belonging to the same UMI-tagged read are then divided into three alignment groups (left, middle and right) according to their relative mapping position to the original read.

4. Define a genuine breakpoint supporting SA by several criteria.

5. Fusion effect inference for oncogenic fusion using snpEff.

6. Genome Visualization of high reliable fusion.


Lastly, outputs a summary table and breakpoint-spanning reads.



## Installation

### 1. Dependencies

- R packages ("plyr", "data.table", "parallel", "dplyr", "tidyr", "ggplot2")


```java

> install.packages(c("plyr", "data.table", "parallel", "dplyr", "tidyr", "ggplot2"))

```

### 2. Installing SplitFusion

```java
git clone https://github.com/Zheng-NGS-Lab/SplitFusion.git

R CMD INSTALL SplitFusion
```


## Run

### 1. Help
```java
python ./SplitFusion/exec/SplitFusion.py -h

usage: SplitFusion.py [-h] --SplitFusionPath SplitFUSIONPATH --R R
                          --hgRef HGREF --bam_path BAM_PATH --sample_id
                          SAMPLE_ID --output OUTPUT [--panel PANEL]
                          [--fusion_library FUSION_LIBRARY] [--step STEP]
                          [--samtools SAMTOOLS] [--bedtools BEDTOOLS]
                          [--java JAVA] [--bwa BWA] [--snpEff SNPEFF]
                          [--snpEff_ref SNPEFF_REF] [--thread THREAD]
                          [--strVarMinStartSite STRVARMINSTARTSITE]
                          [--maxQueryGap MAXQUERYGAP] [--minMQ MINMQ]
                          [--minMapLength MINMAPLENGTH]
                          [--maxOverlap MAXOVERLAP]
                          [--minExclusive MINEXCLUSIVE]

SplitFusion is a fast data analysis pipeline detects gene fusion based on
split reads and/or paired-end reads.

optional arguments:
  -h, --help            show this help message and exit
  --SplitFusionPath SplitFUSIONPATH
                        the path where Split-Fusion pipeline is installed
                        [required]
  --R R                 the path of R [required]
  --perl PERL           the path of perl [required]
  --hgRef HGREF         the path where human genome reference is stored
                        [required]
  --bam_path BAM_PATH   the path where bam or fastq file is stored [required]
  --sample_id SAMPLE_ID
                        the sample name of running [required]
  --output OUTPUT       the path where output is stored [required]
  --panel PANEL         the path where target genes panel file is stored
  --fusion_library FUSION_LIBRARY
                        the path where fusion library file is stored
  --step STEP           the step of running
  --AnnotationMethod ANNOTATIONMETHOD
                        the name of annotation tools (annovar or snpEff, default: annovar)
  --samtools SAMTOOLS   the path of samtools
  --bedtools BEDTOOLS   the path of bedtools
  --java JAVA           the path of java
  --bwa BWA             the path of bwa
  --snpEff SNPEFF       the path of snpEff
  --snpEff_ref SNPEFF_REF
                        the version of snpEff reference
  --thread THREAD       threads of BWA
  --strVarMinStartSite STRVARMINSTARTSITE
                        minimum start site
  --maxQueryGap MAXQUERYGAP
                        maximum gap length
  --minMQ MINMQ         minimum mapping quality
  --minMapLength MINMAPLENGTH
                        minimum read mapping length
  --maxOverlap MAXOVERLAP
                        maximum overlap length
  --minExclusive MINEXCLUSIVE
                        minimum exclusive length

```

### 2. run SplitFusion
```java
python ./SplitFusion/exec/SplitFusion.py --SplitFusionPath SplitFusionPath --hgRef hgRef --bam_path bam_path --sample_id sample_id --output output --R R --perl perl
```

## Output 
[An example brief output table:](https://github.com/Zheng-NGS-Lab/SplitFusion/blob/master/inst/data/example_data/result/example/example.brief.summary)

| AP7         | GeneExon5'---GeneExon3'    | num_unique_reads | frame    | Gene_Exon_cDNA_5'_3'            |
|:-----------:|:--------------------------:|:----------------:|:--------:|:-------------------------------:|
| A01-P701    | KIF5B_exon15---RET_exon12  |                7 | in-frame | KIF5B exon15 c.1723 .NM_004521.---RET exon12 c.2138 .NM_020630. |
| A02-P702    | EML4_intronic---ALK_exon20 |                9 | _NA_     | EML4 intronic c.NA .NM_001145076.---ALK exon20 c.3171 .NM_004304. |
| A02-P702    | EML4_intronic---ALK_exon20 |               10 | _NA_     | EML4 intronic c.NA .NM_001145076.---ALK exon20 c.3173 .NM_004304. |
| A02-P702    | EML4_exon4---ALK_exon20    |               64 | in-frame | EML4 exon4 c.468 .NM_001145076.---ALK exon20 c.3171 .NM_004304. |

[An example output fastq file for the KIF5B_exon15---RET_exon12 fusion of sample A01-P701 is:](https://github.com/Zheng-NGS-Lab/SplitFusion/blob/master/inst/data/example_data/result/example/example.EML4_intron6---ALK_exon20.txt)

 >CL100059760L2C005R002_288074
TTCCCACTTTGGATCCTCCTTTACATCATTATTTCCCACAGCAATTCCTATTTCTGCAAGGTCTTTTAGTAAAGATGC
 >CL100059760L2C008R017_227619
TTCCGAGGGAATTCCCACTTTGGATCCTCCTTTACATCATTATTTCCCACAGCAATTCCTATTTCTGCAAGGTCTTTT
 >CL100059760L2C005R026_187567
TTGACTGGAGTTCAGACGTGTGCTCTTCCGAAAGCCCTCCCCGGTGCGCATGTTGGCAGGCTCAGACAAGGCCCTGG
 >CL100059760L2C003R075_538109
TAGGAATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGAA
 >CL100059760L2C006R047_14778
ATAGGAATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGA
 >CL100059760L2C012R012_483791
GAATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGAAAAA
 >CL100059760L2C013R006_484169
AATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGAAAAAC
 >CL100059760L2C017R020_507274
ATAGGAATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGA
 >CL100059760L2C017R091_315407
GAATTGCTGTGGGAAATAATGATGTAAAGGAGGATCCAAAGTGGGAATTCCCTCGGAAGAGCTTGGTTCTTGGAAAAA
 >CL100059760L2C003R015_252083
GGGAATTCCCACTTTGGATCCTCCTATGTTGGAATTCCCTCGGAAGAACTTGGTTCTTGGAAAAACTCTAAGATCGGA



## Visualization
[An visualization of example output fastq for the EML4_intron6---ALK_exon20:](https://github.com/Zheng-NGS-Lab/SplitFusion/blob/master/inst/data/example_data/result/example/example.EML4_intron6---ALK_exon20.png)
![image](https://github.com/Zheng-NGS-Lab/SplitFusion/blob/master/inst/data/example_data/result/example/example.EML4_intron6---ALK_exon20.png)




